File: src/wwwroot/app/common/models/business/chainTransaction/chainTransaction.ts
/// <reference path="../../../../typings/_all.d.ts" />

module Antares.Common.Models.Business {
    export class ChainTransaction implements Dto.IChainTransaction {
        id: string = '';
        propertyId: string = '';
        mortgageId: string = '';
        surveyId: string = '';
        searchesId: string = '';
        enquiriesId: string = '';
        contractAgreedId: string = '';
        activity: Business.Activity = null;
        requirement: Business.Requirement = null;
        parent: ChainTransaction = null;
        isEnd: boolean = false;
        property: Business.PreviewProperty = null;
        vendor: string = '';
        agentUser: Business.User = null;
        agentContact: Business.Contact = null;
        agentCompany: Business.Company = null;
        agentCompanyContact: Business.CompanyContactRelation = null;
        solicitorContact: Business.Contact = null;
        solicitorCompany: Business.Company = null;
        solicitorCompanyContact: Business.CompanyContactRelation = null;
        mortgage: Business.EnumTypeItem = null;
        survey: Business.EnumTypeItem = null;
        searches: Business.EnumTypeItem = null;
        enquiries: Business.EnumTypeItem = null;
        contractAgreed: Business.EnumTypeItem = null;
        surveyDate: Date | string = null;
        isKnightFrankAgent: string = null;

        constructor(chainTransaction?: Dto.IChainTransaction) {
            angular.extend(this, chainTransaction);
            if (chainTransaction) {
                this.property = new Business.PreviewProperty(chainTransaction.property);
                if (chainTransaction.activity) {
                    this.activity = new Business.Activity(chainTransaction.activity);
                }
                if (chainTransaction.requirement) {
                    this.requirement = new Business.Requirement(chainTransaction.requirement);
                }
                if (chainTransaction.parent) {
                    this.parent = new ChainTransaction(chainTransaction.parent);
                }
                if (chainTransaction.agentUser) {
                    this.agentUser = new User(chainTransaction.agentUser);
                }
                if (chainTransaction.agentContact && chainTransaction.agentCompany) {
                    this.agentContact = new Contact(chainTransaction.agentContact);
                    this.agentCompany = new Company(chainTransaction.agentCompany);
                    this.agentCompanyContact = new CompanyContactRelation(this.agentContact, this.agentCompany);
                }
                if (chainTransaction.solicitorContact && chainTransaction.solicitorCompany) {
                    this.solicitorContact = new Contact(chainTransaction.solicitorContact);
                    this.solicitorCompany = new Company(chainTransaction.solicitorCompany);
                    this.solicitorCompanyContact = new CompanyContactRelation(this.solicitorContact, this.solicitorCompany);
                }
                if (chainTransaction.surveyDate) {
                    this.surveyDate = new Date(<string>chainTransaction.surveyDate);
                }
            }
        }
    }
}

File: src/KnightFrank.Antares.Dal/Model/Offer/ChainTransaction .cs
namespace KnightFrank.Antares.Dal.Model.Offer
{
    using System;

    using Common;
    using Enum;
    using Contacts;

    using KnightFrank.Antares.Dal.Model.Company;
    using KnightFrank.Antares.Dal.Model.Property.Activities;
    using KnightFrank.Antares.Dal.Model.User;
    using Property;

    public class ChainTransaction : BaseEntity, IAuditableEntity
    {
        public Guid? ActivityId { get; set; }

        public virtual Activity Activity { get; set; }

        public Guid? RequirementId { get; set; }

        public virtual Requirement Requirement { get; set; }

        public Guid? ParentId { get; set; }

        public ChainTransaction Parent { get; set; }

        public bool IsEnd { get; set; }

        public Guid PropertyId { get; set; }

        public virtual Property Property { get; set; }

        public string Vendor { get; set; }

        public Guid? AgentUserId { get; set; }

        public virtual User AgentUser { get; set; }

        public Guid? AgentContactId { get; set; }

        public virtual Contact AgentContact { get; set; }

        public Guid? AgentCompanyId { get; set; }

        public virtual Company AgentCompany { get; set; }

        public Guid? SolicitorContactId { get; set; }

        public virtual Contact SolicitorContact { get; set; }

        public Guid? SolicitorCompanyId { get; set; }

        public virtual Company SolicitorCompany { get; set; }

        public Guid MortgageId { get; set; }

        public virtual EnumTypeItem Mortgage { get; set; }

        public Guid SurveyId { get; set; }

        public virtual EnumTypeItem Survey { get; set; }

        public Guid SearchesId { get; set; }

        public virtual EnumTypeItem Searches { get; set; }

        public Guid EnquiriesId { get; set; }

        public virtual EnumTypeItem Enquiries { get; set; }

        public Guid ContractAgreedId { get; set; }

        public virtual EnumTypeItem ContractAgreed { get; set; }

        public DateTime? SurveyDate { get; set; }

        public DateTime CreatedDate { get; set; }

        public DateTime LastModifiedDate { get; set; }

        public bool IsKnightFrankAgent { get; set; }
    }
}
